docs/enhancements/dynamic-fetching-of-openshift-versions.md

```diff
@@ -0,0 +1,280 @@
+---
+title: dynamic-fetching-of-openshift-versions
+authors:
+  - "@osherdp"
+creation-date: 2023-01-12
+last-updated: 2023-01-12
+---
+
+# Dynamic Fetching of OpenShift Versions
+
+Letting users of Assisted Installer API and UI to choose an OpenShift version
+out of a variety of supported versions, all without the need to specifically
+configure each of those versions.
+
+## Summary
+
+Currently, we have defined processes to support the growing list of OpenShift
+versions. It can be separated to the following cases:
+* Adding a newly released minor version of OpenShift when it's first introduced
+  as part of an EC (Engineering Candidate), FC (Feature Complete) or RC (Release
+  Complete) versions. For example, OpenShift 4.12 had its first appearance as
+  version ``4.12.0-ec.0``. With the new version we usually also add live-iso
+  discovery images that are appropriate for the release.
+* Updating beta releases like EC/FC/RC by following candidate channels like
+  ``candidate-4.12``.
+* Updating to the first official stable version when it's GA (Globally
+  Available). For example, updating to 4.12.0 once it's published and available
+  for customers to use.
+* Updating stable releases following stable channels like ``stable-4.12``.
+
+All those actions consist of either manual updates by developers or automatic
+tools like https://gitlab.cee.redhat.com/assisted-installer-ops/assisted-installer-cicd.
+Either manual or automatic, we usually override releases with the new ones (e.g.
+replacing 4.12.0 with 4.12.1) to make sure we don't grow too much our configuration
+files. As per special requests (usually made by customers), we append additional
+versions. For example, see:
+https://gitlab.cee.redhat.com/service/app-interface/-/blob/3fb1ac40a81c7de9bdf1ae6262798fee9336debf/data/services/assisted-installer/cicd/target/production/assisted-service.yaml
+
+The idea of this enhancement is configuring the assisted-service to the OpenShift
+channels it needs to follow (e.g. ``candidate-4.12``, ``stable-4.11``), and let
+it periodically fetch versions to provide them to customers.
+
+## Motivation
+
+It becomes a pretty tedious process to handle each version separately. Some
+use-cases we tend to have:
+* A certain customer use a specific version x.y.z for certification, and we replace
+  this version with a newer one x.y.z+w (w>=1). In this case we'll append again the
+  x.y.z version to the configuration.
+* As described above, we do version updates either automatically or manually.
+* QE team wants to check if a newly added OpenShift version has a regression in its
+  behavior from a previous version.
+* QE team tests a flow which includes OpenShift upgrade process.
+
+### Goals
+
+* Assisted-service is able to provide all versions included in the upgrade-channels,
```

Comment:
* You mean only adding the X latest versions for a channel?
* See below, currently the suggestion is to only include items in this API (and not exclude everything that misses from there). So by default it will always 'lock' a version in the list

Reply:
Yeah, meant we could have a configuration for the number of versions to include. But can be handled later if needed.